pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/BaseRecordExtractor.java

```diff
@@ -0,0 +1,174 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.spi.data.readers;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+
+/**
+ * Base abstract class for extracting and converting the fields of various data formats into supported Pinot data types.
+ *
+ * @param <T> the format of the input record
+ */
+public abstract class BaseRecordExtractor<T> implements RecordExtractor<T> {
+
+  /**
+   * Converts the field value to either a single value (string, number, byte[]), multi value (Object[]) or a Map.
+   * Returns {@code null} if the value is an empty array/collection/map.
+   *
+   * Natively Pinot only understands single values and multi values.
+   * Map is useful only if some ingestion transform functions operates on it in the transformation layer.
+   */
+  @Nullable
+  public Object convert(Object value) {
+    Object convertedValue;
+    if (isInstanceOfMultiValue(value)) {
+      convertedValue = convertMultiValue(value);
+    } else if (isInstanceOfMap(value)) {
+      convertedValue = convertMap(value);
+    } else if (isInstanceOfRecord(value)) {
+      convertedValue = convertRecord(value);
+    } else {
+      convertedValue = convertSingleValue(value);
+    }
+    return convertedValue;
```

Comment:
Suggest renaming some methods:
```suggestion
    if (isMultiValue(value)) {
      return convertMultiValue(value);
    } else if (isMap(value)) {
      return convertMap(value);
    } else if (isRecord(value)) {
      return convertRecord(value);
    } else {
      return convertSingleValue(value);
    }
```